README.md
![Banner](https://s-christy.com/status-banner-service/minesweeper/banner-slim.svg)

## Overview

This is a Minesweeper clone written in C++ using the SDL2 library. It includes
an auto-solving algorithm that not only provides hints for the statistically
best move, but can also automatically solve the game.

## Screenshots

<p align="center">
  <img src="./screenshots/image_0.png" />
</p>

<p align="center">
  <img src="./screenshots/image_1.png" />
</p>

<p align="center">
  <img src="./screenshots/image_2.png" />
</p>

## Features

- Classic Minesweeper gameplay
- Customizable board size and mine count
- Auto-solving algorithm providing hints and automatic solutions
- Statistically based move suggestions for when a guess is required
- Built with C++ and the SDL2 library for cross-platform support
- Command line argument parsing
- Built using CMake
- Image support for mines and flags.
- Font rendering support.
- Error handling.

## Usage

```
Usage:
  -h,--help              Display Usage statement.
  -n,--num-mines         Number of mines
  -x,--width             Board width
  -y,--height            Board height
```

## Controls

- Left-click on a cell to reveal its contents.
- Right-click on a cell to mark it as a mine.
- Press 'h' to get hints for the next best move.
- Press 'c' to let the auto-solver complete the game for you.
- Press 'f' to let the auto-solver mark places where there are definately mines.
- Press 'r' to restart the game with the same settings.
- Press 'ESC' to quit the game.

## Dependencies

```
cmake
g++
libsdl2-dev
libsdl2-gfx-dev
libsdl2-image-dev
libsdl2-ttf-dev
```

## License

This work is licensed under the GNU General Public License version 3 (GPLv3).

[<img src="https://s-christy.com/status-banner-service/GPLv3_Logo.svg" width="150" />](https://www.gnu.org/licenses/gpl-3.0.en.html)


CHANGELOG.md
# Changelog

## [1.0.0] - 2023-04-12

### Added

- Classic Minesweeper gameplay
- Customizable board size and mine count
- Auto-solving algorithm providing hints and automatic solutions
- Statistically based move suggestions for when a guess is required
- Built with C++ and the SDL2 library for cross-platform support
- Command line argument parsing
- Built using CMake
- Image support for mines and flags.
- Font rendering support.
- Error handling.

### Changed

- None (initial release)


src/render.cpp
#include <render.h>

void draw_text(SDL_Renderer *renderer, TTF_Font *font, int x, int y,
               SDL_Color color, const char *text, int square_size) {
  if (!font) {
    return;
  }

  SDL_Surface *textSurface = TTF_RenderText_Solid(font, text, color);
  SDL_Texture *textTexture =
      SDL_CreateTextureFromSurface(renderer, textSurface);
  SDL_Rect rect = textSurface->clip_rect;
  rect.x = x - rect.w / 2 + square_size / 2;
  rect.y = y - rect.h / 2 + square_size / 2;
  SDL_RenderCopy(renderer, textTexture, NULL, &rect);
  SDL_FreeSurface(textSurface);
  SDL_DestroyTexture(textTexture);
}

void render(SDL_Renderer *renderer, Board *board, int mouse_x, int mouse_y,
            TTF_Font *font, SDL_Texture *flag_texture,
            SDL_Texture *mine_texture, int square_size) {
  int shade = 0x07;
  SDL_SetRenderDrawColor(renderer, shade, shade, shade, 0xff);
  SDL_RenderClear(renderer);

  for (int x = 0; x < board->width; x++) {
    for (int y = 0; y < board->height; y++) {
      board->squares[x][y].is_hover = false;
    }
  }

  Square *s = board->GetCollision(mouse_x, mouse_y);
  if (s) {
    s->is_hover = true;
  }

  bool finished = false;
  while (!finished) {
    finished = true;
    for (int x = 0; x < board->width; x++) {
      for (int y = 0; y < board->height; y++) {
        int num_neighbors = board->GetNumNeighbors(x, y);
        Square *square = board->GetSquare(x, y);
        if (square) {
          if (square->state == UNCOVERED && num_neighbors == 0) {
            for (int cx = x - 1; cx <= x + 1; cx++) {
              for (int cy = y - 1; cy <= y + 1; cy++) {
                Square *s = board->GetSquare(cx, cy);
                if (s) {
                  if (s->state == COVERED) {
                    s->state = UNCOVERED;
                    finished = false;
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  for (int x = 0; x < board->width; x++) {
    for (int y = 0; y < board->height; y++) {
      SDL_Rect rect;
      rect.x = x * square_size;
      rect.y = y * square_size;
      rect.w = square_size - 1;
      rect.h = square_size - 1;

      Square *square = board->GetSquare(x, y);
      if (square) {
        if (square->is_hover && square->state == COVERED) {
          int shade = 0x7f;
          SDL_SetRenderDrawColor(renderer, shade, shade, shade, 0xff);
          SDL_RenderFillRect(renderer, &rect);
          if (square->is_flag && flag_texture) {
            SDL_Rect flag_rect;
            flag_rect.x = rect.x;
            flag_rect.y = rect.y;
            flag_rect.w = square_size;
            flag_rect.h = square_size;
            SDL_RenderCopy(renderer, flag_texture, NULL, &flag_rect);
          }
        } else if (square->state == COVERED) {
          int shade = 0xaf;
          SDL_SetRenderDrawColor(renderer, shade, shade, shade, 0xff);
          SDL_RenderFillRect(renderer, &rect);

          SDL_Rect inside_rect;
          inside_rect.x = rect.x + 1;
          inside_rect.y = rect.y + 1;
          inside_rect.w = square_size - 2;
          inside_rect.h = square_size - 2;
          shade = 0x7f;
          SDL_SetRenderDrawColor(renderer, shade, shade, shade, 0xff);
          SDL_RenderFillRect(renderer, &inside_rect);
          if (square->is_flag && flag_texture) {
            SDL_Rect flag_rect;
            flag_rect.x = rect.x;
            flag_rect.y = rect.y;
            flag_rect.w = square_size;
            flag_rect.h = square_size;
            SDL_RenderCopy(renderer, flag_texture, NULL, &flag_rect);
          }
        } else if (square->state == UNCOVERED) {
          if (square->success_probability == 0) {
            int shade = 0xff;
            SDL_SetRenderDrawColor(renderer, shade, shade, shade, 0xff);
            SDL_RenderFillRect(renderer, &rect);
          } else if (square->success_probability <= 34) {
            int shade = 0xaf;
            SDL_SetRenderDrawColor(renderer, 0xff, shade, shade, 0xff);
            SDL_RenderFillRect(renderer, &rect);
          } else if (square->success_probability <= 50) {
            int shade = 0xaf;
            SDL_SetRenderDrawColor(renderer, 0xff, 0xff, shade, 0xff);
            SDL_RenderFillRect(renderer, &rect);
          } else {
            int shade = 0xaf;
            SDL_SetRenderDrawColor(renderer, shade, 0xff, shade, 0xff);
            SDL_RenderFillRect(renderer, &rect);
          }

          SDL_Color black = {0, 0, 0};
          if (square->is_mine && mine_texture) {
            SDL_Rect mine_rect;
            mine_rect.x = rect.x;
            mine_rect.y = rect.y;
            mine_rect.w = square_size;
            mine_rect.h = square_size;
            SDL_RenderCopy(renderer, mine_texture, NULL, &mine_rect);
          } else {
            int num_neighbors = board->GetNumNeighbors(x, y);
            if (num_neighbors > 0 && font) {
              char num_neigbors_text[8];
              sprintf(num_neigbors_text, "%d", num_neighbors);
              draw_text(renderer, font, rect.x, rect.y, black,
                        num_neigbors_text, square_size);
            }
          }
        }
      }
    }
  }

  SDL_RenderPresent(renderer);
}


src/command_line.h
#ifndef COMMAND_LINE_H
#define COMMAND_LINE_H

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct Argument {
  char *longName;
  char *description;
  char *value;
  int set;
  void (*function)();
};

struct Argument g_arguments[26] = {{0}};

void set_arg_function(void (*function)(), char short_name) {
  int idx = short_name - 'a';

  if (idx >= 0 && idx < 26) {
    g_arguments[idx].function = function;
  }
}

void add_arg(char short_name, const char *longName, const char *description) {
  int idx = short_name - 'a';

  if (idx >= 0 && idx < 26) {
    g_arguments[idx].longName = (char *)malloc(strlen(longName) + 1);
    g_arguments[idx].description = (char *)malloc(strlen(description) + 1);

    strcpy(g_arguments[idx].longName, longName);
    strcpy(g_arguments[idx].description, description);
  }
}

int get_is_set(char short_name) {
  int idx = short_name - 'a';

  if (idx >= 0 && idx < 26) {
    return g_arguments[idx].set;
  }

  return 0;
}

char *get_value(char short_name) {
  int idx = short_name - 'a';

  if (idx >= 0 && idx < 26) {
    if (g_arguments[idx].value) {
      return g_arguments[idx].value;
    }
  }

  return (char *)malloc(0);
}

void usage() {
  printf("Usage:\n");
  for (int i = 0; i < 26; i++) {
    if (g_arguments[i].description) {
      printf("  -%c,--%s ", 'a' + i, g_arguments[i].longName);
      for (int j = strlen(g_arguments[i].longName); j < 17; j++) {
        printf(" ");
      }
      printf("%s\n", g_arguments[i].description);
    }
  }

  exit(EXIT_SUCCESS);
}

void parse_opts(int argc, char *argv[]) {
  add_arg('h', "help", "Display Usage statement.");
  set_arg_function(usage, 'h');

  int last_arg = -1;

  for (int i = 1; i < argc; i++) {
    if (strlen(argv[i]) >= 2) {
      if (argv[i][0] == '-') {

        // Short opts
        if (argv[i][1] != '-') {
          for (int j = 1; j < strlen(argv[i]); j++) {
            if (argv[i][j] >= 'a' && argv[i][j] <= 'z') {
              last_arg = argv[i][j] - 'a';
              g_arguments[last_arg].set = 1;
              if (g_arguments[last_arg].function) {
                g_arguments[last_arg].function();
              }
            }
          }

          // Long opts
        } else {
          if (strlen(argv[i]) >= 3) {
            for (int j = 0; j < 26; j++) {
              if (g_arguments[j].longName) {
                if (strcmp(g_arguments[j].longName, argv[i] + 2) == 0) {
                  last_arg = j;
                  g_arguments[last_arg].set = 1;
                  if (g_arguments[last_arg].function) {
                    g_arguments[last_arg].function();
                  }
                  break;
                }
              }
            }

            // Double hyphen
          } else {
            return;
          }
        }

        // Value
      } else {
        if (last_arg != -1) {
          g_arguments[last_arg].value = (char *)malloc(strlen(argv[i]) + 1);
          strcpy(g_arguments[last_arg].value, argv[i]);
        }
      }

      // Single character arg
    } else {
      if (last_arg != -1) {
        g_arguments[last_arg].value = (char *)malloc(strlen(argv[i]) + 1);
        strcpy(g_arguments[last_arg].value, argv[i]);
      }
    }
  }
}

#endif


src/main.cpp
#include <SDL2/SDL.h>
#include <SDL2/SDL2_gfxPrimitives.h>
#include <SDL2/SDL_image.h>
#include <SDL2/SDL_ttf.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <command_line.h>
#include <render.h>

#define FLAG_PATH "res/flag.bmp"
#define FONT_PATH "res/LiberationSans-Regular.ttf"
#define MINE_PATH "res/mine.bmp"
#define SMALL_FONT_PATH "res/LiberationSans-Regular.ttf"
#define SQUARE_SIZE 40

bool rand_bool(float p) { return (float)rand() / (float)RAND_MAX < p; }

Square::Square() { this->Initialize(); }

void Square::Initialize() {
  this->state = COVERED;
  this->is_hover = false;
  this->is_mine = false;
  this->is_flag = false;
  this->success_probability = 0;
}

void Board::SeedRandomMine() {
  int num_unmined_squares = 0;

  for (int x = 0; x < this->width; x++) {
    for (int y = 0; y < this->height; y++) {
      if (!this->GetSquare(x, y)->is_mine) {
        num_unmined_squares++;
      }
    }
  }

  if (num_unmined_squares == 0) {
    return;
  }

  int choice = random() % num_unmined_squares;

  int counter = 0;
  for (int x = 0; x < this->width; x++) {
    for (int y = 0; y < this->height; y++) {
      if (!this->GetSquare(x, y)->is_mine) {
        if (counter == choice) {
          this->GetSquare(x, y)->is_mine = true;
          static int nm = 0;
          nm++;
          return;
        }
        counter++;
      }
    }
  }
}

void Board::AutoFlag() {
  for (int x = 0; x < this->width; x++) {
    for (int y = 0; y < this->height; y++) {
      Square *s = this->GetSquare(x, y);
      if (s) {
        if (s->state == UNCOVERED) {
          int num_neighbors = this->GetNumNeighbors(x, y);
          int num_neighbors_covered = this->GetNumNeighborsCovered(x, y);
          if (num_neighbors == num_neighbors_covered) {
            for (int cx = x - 1; cx <= x + 1; cx++) {
              for (int cy = y - 1; cy <= y + 1; cy++) {
                Square *s = this->GetSquare(cx, cy);
                if (s) {
                  if (s->state == COVERED) {
                    s->is_flag = true;
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

void Board::MarkProbability() {
  for (int x = 0; x < this->width; x++) {
    for (int y = 0; y < this->height; y++) {
      Square *s = this->GetSquare(x, y);
      if (s) {
        s->success_probability = 0;
        int num_neighbors = this->GetNumNeighbors(x, y);
        int num_neighbors_flagged = this->GetNumNeighborsFlagged(x, y);
        int num_neighbors_covered = this->GetNumNeighborsCovered(x, y);
        if (s->state == UNCOVERED && num_neighbors != 0 &&
            num_neighbors_flagged != num_neighbors_covered) {
          int choices = num_neighbors_covered - num_neighbors_flagged;
          int dangerous = num_neighbors - num_neighbors_flagged;
          s->success_probability = 100 - 100 * dangerous / choices;
        }
      }
    }
  }
}

void Board::AutoComplete() {
  bool finished = false;
  while (!finished) {
    finished = true;
    this->AutoFlag();
    for (int x = 0; x < this->width; x++) {
      for (int y = 0; y < this->height; y++) {
        Square *s = this->GetSquare(x, y);
        if (s) {
          if (s->state == UNCOVERED) {
            int num_neighbors = this->GetNumNeighbors(x, y);
            int num_neighbors_flagged = this->GetNumNeighborsFlagged(x, y);
            if (num_neighbors == num_neighbors_flagged) {
              for (int cx = x - 1; cx <= x + 1; cx++) {
                for (int cy = y - 1; cy <= y + 1; cy++) {
                  Square *s = this->GetSquare(cx, cy);
                  if (s) {
                    if (s->state == COVERED && s->is_flag == false) {
                      s->state = UNCOVERED;
                      finished = false;
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  this->MarkProbability();
}

void Board::Reset() {
  for (int x = 0; x < this->width; x++) {
    for (int y = 0; y < this->height; y++) {
      Square *s = this->GetSquare(x, y);
      if (s) {
        s->Initialize();
      }
    }
  }

  for (int i = 0; i < this->num_mines; i++) {
    this->SeedRandomMine();
  }
}

Board::Board(int width, int height, int num_mines) {
  this->width = width;
  this->height = height;
  this->num_mines = num_mines;

  this->squares = new Square *[width];
  for (int i = 0; i < width; i++) {
    this->squares[i] = new Square[height];
  }

  this->Reset();
}

Square *Board::GetCollision(int mx, int my) {
  for (int x = 0; x < this->width; x++) {
    for (int y = 0; y < this->height; y++) {
      SDL_Rect rect;
      rect.x = x * SQUARE_SIZE;
      rect.y = y * SQUARE_SIZE;
      rect.w = SQUARE_SIZE + 1;
      rect.h = SQUARE_SIZE + 1;
      if (mx > rect.x && mx < rect.x + rect.w && my > rect.y &&
          my < rect.y + rect.h) {
        return &this->squares[x][y];
      }
    }
  }

  return nullptr;
}

Square *Board::GetSquare(int x, int y) {
  if (x >= 0 && x < this->width) {
    if (y >= 0 && y < this->height) {
      return &this->squares[x][y];
    }
  }
  return nullptr;
}

int Board::GetNumNeighborsCovered(int x, int y) {
  int n = 0;
  for (int cx = x - 1; cx <= x + 1; cx++) {
    for (int cy = y - 1; cy <= y + 1; cy++) {
      Square *s = this->GetSquare(cx, cy);
      if (s) {
        if (s->state == COVERED) {
          n++;
        }
      }
    }
  }

  return n;
}

int Board::GetNumNeighborsFlagged(int x, int y) {
  int n = 0;
  for (int cx = x - 1; cx <= x + 1; cx++) {
    for (int cy = y - 1; cy <= y + 1; cy++) {
      Square *s = this->GetSquare(cx, cy);
      if (s) {
        if (s->is_flag == true) {
          n++;
        }
      }
    }
  }

  return n;
}

int Board::GetNumNeighbors(int x, int y) {
  int n = 0;
  for (int cx = x - 1; cx <= x + 1; cx++) {
    for (int cy = y - 1; cy <= y + 1; cy++) {
      Square *s = this->GetSquare(cx, cy);
      if (s) {
        if (s->is_mine) {
          n++;
        }
      }
    }
  }

  return n;
}

bool event_loop(Board *board, int *mouse_x, int *mouse_y) {
  static SDL_Event event;

  bool running = true;

  int clicked_x = -1;
  int clicked_y = -1;
  int clicked_button = -1;

  Uint32 begin = SDL_GetTicks();

  while (SDL_PollEvent(&event)) {
    switch (event.type) {

    case SDL_QUIT:
      running = false;
      break;

    case SDL_MOUSEMOTION:
      *mouse_x = event.motion.x;
      *mouse_y = event.motion.y;
      break;

    case SDL_MOUSEBUTTONDOWN:
      clicked_x = event.button.x;
      clicked_y = event.button.y;
      clicked_button = event.button.button;
      break;

    case SDL_KEYDOWN:

      switch (event.key.keysym.sym) {

      case SDLK_ESCAPE:
        running = false;
        break;

      case SDLK_r:
        board->Reset();
        break;

      case SDLK_f:
        board->AutoFlag();
        break;

      case SDLK_h:
        board->MarkProbability();
        break;

      case SDLK_c:
        board->AutoComplete();
        break;

      default:
        // printf("Unhandled Key: %d\n", event.key.keysym.sym);
        break;
      }
      break;

    case SDL_POLLSENTINEL:
      break;

    default:
      // printf("Unhandled Event: %d\n", event.type);
      break;
    }
  }

  if (clicked_x != -1) {
    Square *s = board->GetCollision(clicked_x, clicked_y);
    if (s) {
      if (clicked_button == SDL_BUTTON_LEFT) {
        s->state = UNCOVERED;
      } else if (clicked_button == SDL_BUTTON_RIGHT) {
        s->is_flag = !s->is_flag;
      }
    }
  }

  return running;
}

int main(int argc, char *argv[]) {
  srand(time(0));

  int mouse_x = 0;
  int mouse_y = 0;

  add_arg('n', "num-mines", "Number of mines");
  add_arg('x', "width", "Board width");
  add_arg('y', "height", "Board height");

  parse_opts(argc, argv);

  int width = 0;
  int height = 0;
  int num_mines = 0;

  if (get_is_set('x')) {
    width = atoi(get_value('x'));
  }

  if (get_is_set('y')) {
    height = atoi(get_value('y'));
  }

  if (get_is_set('n')) {
    num_mines = atoi(get_value('n'));
  }

  if (!width) {
    width = 30;
  }

  if (!height) {
    height = 15;
  }

  if (!num_mines) {
    num_mines = 15;
  }

  SDL_Init(SDL_INIT_VIDEO | SDL_INIT_EVENTS);
  TTF_Init();

  TTF_Font *font = TTF_OpenFont(FONT_PATH, SQUARE_SIZE);
  TTF_Font *small_font = TTF_OpenFont(SMALL_FONT_PATH, SQUARE_SIZE / 2);

  SDL_Window *window = SDL_CreateWindow(
      "Minesweeper", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED,
      width * SQUARE_SIZE, height * SQUARE_SIZE, SDL_WINDOW_SHOWN);

  SDL_Renderer *renderer =
      SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED);

  Board *board = new Board(width, height, num_mines);

  SDL_Texture *flag_texture = IMG_LoadTexture(renderer, FLAG_PATH);
  SDL_Texture *mine_texture = IMG_LoadTexture(renderer, MINE_PATH);

  if (!font) {
    fprintf(stderr, "Resource: \"%s\" could not be loaded...\n", FONT_PATH);
  }

  if (!small_font) {
    fprintf(stderr, "Resource: \"%s\" could not be loaded...\n",
            SMALL_FONT_PATH);
  }

  if (!flag_texture) {
    fprintf(stderr, "Resource: \"%s\" could not be loaded...\n", FLAG_PATH);
  }

  if (!mine_texture) {
    fprintf(stderr, "Resource: \"%s\" could not be loaded...\n", MINE_PATH);
  }

  int shade = 0x07;
  SDL_SetRenderDrawColor(renderer, shade, shade, shade, 0xff);
  SDL_RenderClear(renderer);

  bool running = true;
  while (running) {
    running = event_loop(board, &mouse_x, &mouse_y);

    bool game_complete = true;
    for (int x = 0; x < board->width; x++) {
      for (int y = 0; y < board->height; y++) {
        Square *s = board->GetSquare(x, y);
        if (!s->is_mine && s->state == COVERED) {
          game_complete = false;
        }
      }
    }

    if (game_complete) {
      board->AutoFlag();
    }

    render(renderer, board, mouse_x, mouse_y, font, flag_texture, mine_texture, SQUARE_SIZE);
  }

  SDL_DestroyRenderer(renderer);
  SDL_DestroyWindow(window);
  SDL_Quit();
}
